packages/go_router/lib/src/delegate.dart

```diff
@@ -165,8 +176,15 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   /// See also:
   /// * [push] which pushes the given location onto the page stack.
   void replace(RouteMatchList matches) {
-    _matchList.pop();
-    push(matches); // [push] will notify the listeners.
+    assert(matches.last.route is! ShellRoute);
+    PageKey pageKey = _matchList.pop().pageKey;
+    if (pageKey.path != matches.fullpath) {
```

Comment:
Seems weird to sometimes reuse and sometimes push. I am not 100% sure this is the route we should go. I was looking at the navigator API, it looks like this mixes both Navigator.replace and Navigator.pushReplacement. Based on the implemented API, this is more close to `pushReplacement`. In that case we should probably always trigger a push, and create another API to do the `replace`

Reply:
Are you suggesting that we implement a
- `pushReplacement` that always uses a new key 
- `replace` that always reuses the key of the replaced page

?

If yes, how would you want me to name those methods? `pushReplacement` and `replace` too? I don't know whether it will be considered as a breaking change or not

And in the case of `/a` being replaced by `/b` with `replace`, would it make sense to reuse `/a`s key and see `/b`'s screen while using `/a` as a key?

Reply:
I'm okay to rename the current `replace` into `pushReplacement`.
But how do I handle that? Do I make it a breaking change? Or do I deprecate `replace` and make it return `pushReplacement` ? 

Also, I guess you want this `pushReplacement` to already generate and use a brand new key, right?
The fix needed to fix https://github.com/flutter/flutter/issues/115902 is to reuse the key when possible

Reply:
I will leave the original issue as a feature request, but I don't think that is something we should do.

Reply:
Do you mean you want to leave `replace` (and reuse the same key) as a feature request but it would not be implemented? 

How would suggest I can fix the bug mentioned in https://github.com/flutter/flutter/issues/115902?

Reply:
I can see a lot of corner cases if we implement this API, if you have an idea on how to implement it, feel free to send a pr

Reply:
What about something similar to what I implemented in this PR?
It would reuse the same key if the path (ex: `/families/:fid`) is the same. Otherwise, it creates a new key?

I can modify this PR once https://github.com/flutter/packages/pull/2848 is merged